File: 0090.Subsets_II.java
//1ms 99.84%RT 20250314

class Solution {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);  
        backtrack(new ArrayList<>(), 0, nums);
        return res;
    }

    public void backtrack(List<Integer> path, int st, int[] nums){
        res.add(new ArrayList<>(path));  

        for (int i = st; i < nums.length; i++){
            if (i > st && nums[i] == nums[i - 1]){
                continue;  
            }
            path.add(nums[i]);
            backtrack(path, i + 1, nums);
            path.remove(path.size() - 1); 
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {1, 2, 2};
        List<List<Integer>> result = solution.subsetsWithDup(nums);
        for (List<Integer> subset : result) {
            System.out.println(subset);
        }
    }
}

File: 0090.Subsets_II.rs
//0ms


use std::collections::HashSet;

impl Solution {
    pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = nums.clone(); // Clone nums to avoid mutating the input directly
        nums.sort(); // Sort the array to handle duplicates

        let n = nums.len(); 
        let mut result = Vec::new();
        let power_of_two = 1 << n; // 2 ^ n

        for i in 0..power_of_two {
            let mut subset = Vec::new();

            for j in 0..n {
                let currentBit = (i >> j) & 1;

                if currentBit == 1 {
                    subset.push(nums[j])
                }
            }

            result.push(subset);
        }

        return Self::remove_duplicates(result)
    }

    fn remove_duplicates(array: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut set = HashSet::new();

        for subset in array {
            set.insert(subset);
        }

        let unique_subsets: Vec<Vec<i32>> = set.into_iter().collect();
        unique_subsets
    }
}
